catboost/libs/fstr/shap_values.cpp

```diff
@@ -141,6 +143,117 @@ static void UpdateShapByFeaturePath(
     }
 }
 
+static TContribution SumContributions(
+    const TContribution& lhs,
+    const TContribution& rhs
+) {
+    CB_ENSURE_INTERNAL(
+        lhs.PositiveContribution.size() == rhs.PositiveContribution.size(), 
+        "Contributions have different sizes");
+    TContribution result{lhs.PositiveContribution.size()};
+    const auto approxDimension = result.PositiveContribution.size();
+    for (size_t dimension = 0; dimension < approxDimension; ++dimension) {
+        result.PositiveContribution[dimension] = lhs.PositiveContribution[dimension] + rhs.PositiveContribution[dimension];
+        result.NegativeContribution[dimension] = lhs.NegativeContribution[dimension] + rhs.NegativeContribution[dimension];
+    }
+    return result;
+}
+
+ // Independent Tree SHAP
+TContribution TInternalIndependentTreeShapCalcer::Calc(
+    int depth,
+    size_t nodeIdx,
+    ui32 uniqueFeaturesCount,
+    ui32 featureMatchedDatasetSampleCount
+) {
+    const auto approxDimension = Forest.GetDimensionsCount();
+    const double* leafValuesPtr = Forest.GetFirstLeafPtrForTree(TreeIdx);
+    if (depth == Forest.GetTreeSizes()[TreeIdx]) {
+        TContribution contribution{approxDimension}; // (pos, neg) = (0, 0)
+        if (featureMatchedDatasetSampleCount == 0) {
+            for (size_t dimension = 0; dimension < approxDimension; ++dimension) {
+                double value = leafValuesPtr[nodeIdx * approxDimension + dimension];
+                ShapValuesInternalByDepth[depth][dimension] += value;
+            }
+        }
+        if (uniqueFeaturesCount != 0) {
+            if (featureMatchedDatasetSampleCount != 0) {
+                for (size_t dimension = 0; dimension < approxDimension; ++dimension) {
+                    contribution.PositiveContribution[dimension] =
+                        Weights[featureMatchedDatasetSampleCount - 1][uniqueFeaturesCount]
+                        * leafValuesPtr[nodeIdx * approxDimension + dimension]; 
+                }
+            }
+            if (featureMatchedDatasetSampleCount != uniqueFeaturesCount) {
+                for (size_t dimension = 0; dimension < approxDimension; ++dimension) {
+                    contribution.NegativeContribution[dimension] =
+                        -Weights[featureMatchedDatasetSampleCount][uniqueFeaturesCount]
+                        * leafValuesPtr[nodeIdx * approxDimension + dimension];
+                }
+            }
+        }
+        return contribution;
+    }
+    const size_t remainingDepth = Forest.GetTreeSizes()[TreeIdx] - depth - 1;
+    const int combinationClass = BinFeatureCombinationClass[
+        Forest.GetTreeSplits()[Forest.GetTreeStartOffsets()[TreeIdx] + remainingDepth]
+    ];
```

Comment:
for faster performance, fill BinFeatureCombinationClass for all possible depths in constructor